File: server/src/match.ts
// The PURE FFA deathmatch reducer. No I/O, no Rapier, no ws, no DOM — `stepMatch`
// takes the current state plus a per-tick aggregate snapshot and mutates the
// state, returning the events the Room must act on (match start, match end,
// reset). Frag COUNTS are NOT stored here — they live on RoomPlayer and are
// incremented at the kill site; the reducer only reads the aggregate (top frags,
// connected count) so it stays pure and exhaustively unit-testable.
//
// Phases: warmup (<2 players, free-roam) -> live (>=2, clock counts up) ->
// matchEnd (scoreboard for matchEndSec) -> reset -> warmup. Timings come from
// TUNING.world so the reducer never drifts from the single source of truth.

import { TUNING } from '@rivals/shared';

export type MatchPhase = 'warmup' | 'live' | 'matchEnd';

export interface MatchState {
  phase: MatchPhase;
  clock: number; // live: seconds elapsed in the match; matchEnd: seconds left on display
  matchWinner: number; // playerId at matchEnd, else -1
}

export interface MatchTickCtx {
  connectedCount: number;
  topFrags: number; // highest frag count among players this tick
  topFragsPlayer: number; // that player's id (lowest id on ties), -1 if none
}

export type MatchEvent =
  | { type: 'matchStart' }
  | { type: 'matchEnd'; winner: number }
  | { type: 'reset' };

const W = TUNING.world;

export function initMatch(): MatchState {
  return { phase: 'warmup', clock: 0, matchWinner: -1 };
}

/**
 * Advance the match by `dt` seconds against this tick's aggregate context.
 * MUTATES `state` and returns the events produced this tick (0 or 1). The Room
 * calls this once per server tick.
 */
export function stepMatch(state: MatchState, ctx: MatchTickCtx, dt: number): MatchEvent[] {
  const events: MatchEvent[] = [];

  switch (state.phase) {
    case 'warmup': {
      if (ctx.connectedCount >= W.warmupMinPlayers) {
        state.phase = 'live';
        state.clock = 0;
        state.matchWinner = -1;
        events.push({ type: 'matchStart' });
      }
      break;
    }

    case 'live': {
      state.clock += dt;
      if (ctx.connectedCount < W.warmupMinPlayers) {
        // Not enough players to keep playing — pause back to warmup. Frags are
        // kept on the players; only an actual matchEnd->reset zeroes them.
        state.phase = 'warmup';
        break;
      }
      const reachedLimit = ctx.topFragsPlayer >= 0 && ctx.topFrags >= W.fragLimit;
      const reachedCap = W.matchTimeCapSec > 0 && state.clock > W.matchTimeCapSec;
      if (reachedLimit || reachedCap) {
        state.matchWinner = ctx.topFragsPlayer;
        state.phase = 'matchEnd';
        state.clock = W.matchEndSec;
        events.push({ type: 'matchEnd', winner: ctx.topFragsPlayer });
      }
      break;
    }

    case 'matchEnd': {
      state.clock -= dt;
      if (state.clock <= 0) {
        state.phase = 'warmup';
        state.clock = 0;
        state.matchWinner = -1;
        events.push({ type: 'reset' });
      }
      break;
    }
  }

  return events;
}

File: godot/src/Server/Match.cs
// The PURE FFA deathmatch reducer. No I/O, no engine, no networking. Ported from
// server/src/match.ts. StepMatch takes the current state plus a per-tick aggregate
// and mutates the state, returning the events the Room must act on.
//
// Phases: warmup (<2 players) -> live (>=2, clock counts up) -> matchEnd
// (scoreboard for matchEndSec) -> reset -> warmup.

using System.Collections.Generic;
using WebRivals.Shared;

namespace WebRivals.Server;

public sealed class MatchState
{
    public string Phase = "warmup"; // "warmup" | "live" | "matchEnd"
    public double Clock = 0;
    public int MatchWinner = -1;
}

public struct MatchTickCtx
{
    public int ConnectedCount;
    public int TopFrags;
    public int TopFragsPlayer; // lowest id on ties, -1 if none
}

public enum MatchEventType { MatchStart, MatchEnd, Reset }

public struct MatchEvent
{
    public MatchEventType Type;
    public int Winner; // for MatchEnd
}

public static class Match
{
    private static WorldTuning W => Tuning.World;

    public static MatchState InitMatch() => new MatchState { Phase = "warmup", Clock = 0, MatchWinner = -1 };

    /// <summary>Advance the match by dt against this tick's aggregate. Mutates
    /// state and returns the events produced this tick (0 or 1).</summary>
    public static List<MatchEvent> StepMatch(MatchState state, MatchTickCtx ctx, double dt)
    {
        var events = new List<MatchEvent>();

        switch (state.Phase)
        {
            case "warmup":
                if (ctx.ConnectedCount >= W.WarmupMinPlayers)
                {
                    state.Phase = "live";
                    state.Clock = 0;
                    state.MatchWinner = -1;
                    events.Add(new MatchEvent { Type = MatchEventType.MatchStart });
                }
                break;

            case "live":
                state.Clock += dt;
                if (ctx.ConnectedCount < W.WarmupMinPlayers)
                {
                    state.Phase = "warmup";
                    break;
                }
                bool reachedLimit = ctx.TopFragsPlayer >= 0 && ctx.TopFrags >= W.FragLimit;
                bool reachedCap = W.MatchTimeCapSec > 0 && state.Clock > W.MatchTimeCapSec;
                if (reachedLimit || reachedCap)
                {
                    state.MatchWinner = ctx.TopFragsPlayer;
                    state.Phase = "matchEnd";
                    state.Clock = W.MatchEndSec;
                    events.Add(new MatchEvent { Type = MatchEventType.MatchEnd, Winner = ctx.TopFragsPlayer });
                }
                break;

            case "matchEnd":
                state.Clock -= dt;
                if (state.Clock <= 0)
                {
                    state.Phase = "warmup";
                    state.Clock = 0;
                    state.MatchWinner = -1;
                    events.Add(new MatchEvent { Type = MatchEventType.Reset });
                }
                break;
        }

        return events;
    }
}
